pstoreds/addr_book.go

```diff
@@ -2,398 +2,608 @@ package pstoreds
 	logging "github.com/ipfs/go-log"
 
-	ma "github.com/multiformats/go-multiaddr"
-	mh "github.com/multiformats/go-multihash"
-
 	peer "github.com/libp2p/go-libp2p-peer"
 	pstore "github.com/libp2p/go-libp2p-peerstore"
+	pb "github.com/libp2p/go-libp2p-peerstore/pb"
 	pstoremem "github.com/libp2p/go-libp2p-peerstore/pstoremem"
-)
 
-var (
-	log = logging.Logger("peerstore/ds")
-	// The maximum representable value in time.Time is time.Unix(1<<63-62135596801, 999999999).
-	// But it's too brittle and implementation-dependent, so we prefer to use 1<<62, which is in the
-	// year 146138514283. We're safe.
-	maxTime = time.Unix(1<<62, 0)
-
-	ErrTTLDatastore = errors.New("datastore must provide TTL support")
+	ma "github.com/multiformats/go-multiaddr"
+	lru "github.com/raulk/golang-lru"
+	b32 "github.com/whyrusleeping/base32"
 )
 
-// Peer addresses are stored under the following db key pattern:
-// /peers/addr/<b32 peer id no padding>/<hash of maddr>
-var abBase = ds.NewKey("/peers/addrs")
-
-var _ pstore.AddrBook = (*dsAddrBook)(nil)
-
-// dsAddrBook is an address book backed by a Datastore with both an
-// in-memory TTL manager and an in-memory address stream manager.
-type dsAddrBook struct {
-	cache        cache
-	ds           ds.TxnDatastore
-	subsManager  *pstoremem.AddrSubManager
-	writeRetries int
-}
-
 type ttlWriteMode int
 
 const (
 	ttlOverride ttlWriteMode = iota
 	ttlExtend
 )
 
-type cacheEntry struct {
-	expiration time.Time
-	addrs      []ma.Multiaddr
-}
+var (
+	log = logging.Logger("peerstore/ds")
+	// Peer addresses are stored under the following db key pattern:
+	// /peers/addrs/<b32 peer id no padding>
+	addrBookBase = ds.NewKey("/peers/addrs")
+	// GC lookahead entries are stored in keys with pattern:
+	// /peers/gc/addrs/<unix timestamp of next visit>/<peer ID b32> => nil
+	gcLookaheadBase = ds.NewKey("/peers/gc/addrs")
+	arPool          = &sync.Pool{
+		New: func() interface{} {
+			return &addrsRecord{
+				AddrBookRecord: &pb.AddrBookRecord{},
+				dirty:          false,
+			}
+		},
+	}
+)
 
-type addrRecord struct {
-	ttl  time.Duration
-	addr ma.Multiaddr
+// addrsRecord decorates the AddrBookRecord with locks and metadata.
+type addrsRecord struct {
+	sync.RWMutex
+	*pb.AddrBookRecord
+	dirty bool
 }
 
-func (ar *addrRecord) MarshalBinary() ([]byte, error) {
-	ttlB := make([]byte, 8)
-	binary.LittleEndian.PutUint64(ttlB, uint64(ar.ttl))
-	return append(ttlB, ar.addr.Bytes()...), nil
+func (r *addrsRecord) Reset() {
+	r.AddrBookRecord.Reset()
+	r.dirty = false
 }
 
-func (ar *addrRecord) UnmarshalBinary(b []byte) error {
-	ar.ttl = time.Duration(binary.LittleEndian.Uint64(b))
-	// this had been serialized by us, no need to check for errors
-	ar.addr, _ = ma.NewMultiaddrBytes(b[8:])
+// FlushInTxn writes the record to the datastore by calling ds.Put, unless the record is
+// marked for deletion, in which case the deletion is executed via ds.Delete.
+func (r *addrsRecord) FlushInTxn(txn ds.Txn) (err error) {
```

Comment:
Do we need *transaction* semantics or is a simple `Batch` sufficient?

Reply:
Batch is sufficient IMO, good to have your confirmation in that regard. Will switch to batch semantics.

Reply:
Interesting thing. The `Batch` interface only supports write operations (understandably).

Since in badger everything is a transaction, grouping writes in a Batch (which translates to a transaction in Badger), while performing reads outside, will end up creating lots of tiny transactions for each read.

While we _strictly_ don't need transactionality, in practice we're more efficient like this, at least with badger. Any thoughts?

Reply:
@Stebalien you're right. Removing explicit transactions not only doesn't impair performance, it actually improves it at no cost to allocs or memory. My mind was set on standalone DB, where grouping tiny writes in TXs tends to reduce RTT and improve overall performance.

```
→ benchcmp {2b,3b}.txt
benchmark                                                                     old ns/op     new ns/op     delta
BenchmarkDsPeerstore/Badger/AddAddrs-1Addrs-Caching-8                         179696        163472        -9.03%
BenchmarkDsPeerstore/Badger/AddAddrs-10Addrs-Caching-8                        205406        197130        -4.03%
BenchmarkDsPeerstore/Badger/AddAddrs-100Addrs-Caching-8                       367999        363332        -1.27%
BenchmarkDsPeerstore/Badger/AddGetAndClearAddrs-1Addrs-Caching-8              269525        254972        -5.40%
BenchmarkDsPeerstore/Badger/AddGetAndClearAddrs-10Addrs-Caching-8             299234        295765        -1.16%
BenchmarkDsPeerstore/Badger/AddGetAndClearAddrs-100Addrs-Caching-8            462520        404654        -12.51%
BenchmarkDsPeerstore/Badger/Get1000PeersWithAddrs-1Addrs-Caching-8            3204034       3160450       -1.36%
BenchmarkDsPeerstore/Badger/Get1000PeersWithAddrs-10Addrs-Caching-8           3069343       3085779       +0.54%
BenchmarkDsPeerstore/Badger/Get1000PeersWithAddrs-100Addrs-Caching-8          3082513       3106668       +0.78%
BenchmarkDsPeerstore/Badger/GetAddrs-1Addrs-Caching-8                         338           327           -3.25%
BenchmarkDsPeerstore/Badger/GetAddrs-10Addrs-Caching-8                        411           404           -1.70%
BenchmarkDsPeerstore/Badger/GetAddrs-100Addrs-Caching-8                       1065          1071          +0.56%
BenchmarkDsPeerstore/Badger/SetAddrs-1Addrs-Caching-8                         183677        175424        -4.49%
BenchmarkDsPeerstore/Badger/SetAddrs-10Addrs-Caching-8                        206536        205483        -0.51%
BenchmarkDsPeerstore/Badger/SetAddrs-100Addrs-Caching-8                       350174        343417        -1.93%
BenchmarkDsPeerstore/Badger#01/AddAddrs-1Addrs-Cacheless-8                    177433        174883        -1.44%
BenchmarkDsPeerstore/Badger#01/AddAddrs-10Addrs-Cacheless-8                   218380        185580        -15.02%
BenchmarkDsPeerstore/Badger#01/AddAddrs-100Addrs-Cacheless-8                  351320        345705        -1.60%
BenchmarkDsPeerstore/Badger#01/AddGetAndClearAddrs-1Addrs-Cacheless-8         354761        339836        -4.21%
BenchmarkDsPeerstore/Badger#01/AddGetAndClearAddrs-10Addrs-Cacheless-8        410234        398477        -2.87%
BenchmarkDsPeerstore/Badger#01/AddGetAndClearAddrs-100Addrs-Cacheless-8       528694        535091        +1.21%
BenchmarkDsPeerstore/Badger#01/Get1000PeersWithAddrs-1Addrs-Cacheless-8       3255647       3269187       +0.42%
BenchmarkDsPeerstore/Badger#01/Get1000PeersWithAddrs-10Addrs-Cacheless-8      3027279       3079264       +1.72%
BenchmarkDsPeerstore/Badger#01/Get1000PeersWithAddrs-100Addrs-Cacheless-8     3070356       3037135       -1.08%
BenchmarkDsPeerstore/Badger#01/GetAddrs-1Addrs-Cacheless-8                    3225          3159          -2.05%
BenchmarkDsPeerstore/Badger#01/GetAddrs-10Addrs-Cacheless-8                   9312          9423          +1.19%
BenchmarkDsPeerstore/Badger#01/GetAddrs-100Addrs-Cacheless-8                  56322         55458         -1.53%
BenchmarkDsPeerstore/Badger#01/SetAddrs-1Addrs-Cacheless-8                    168983        172715        +2.21%
BenchmarkDsPeerstore/Badger#01/SetAddrs-10Addrs-Cacheless-8                   200770        191014        -4.86%
BenchmarkDsPeerstore/Badger#01/SetAddrs-100Addrs-Cacheless-8                  348456        348516        +0.02%

benchmark                                                                     old allocs     new allocs     delta
BenchmarkDsPeerstore/Badger/AddAddrs-1Addrs-Caching-8                         146            138            -5.48%
BenchmarkDsPeerstore/Badger/AddAddrs-10Addrs-Caching-8                        350            350            +0.00%
BenchmarkDsPeerstore/Badger/AddAddrs-100Addrs-Caching-8                       2385           2386           +0.04%
BenchmarkDsPeerstore/Badger/AddGetAndClearAddrs-1Addrs-Caching-8              214            214            +0.00%
BenchmarkDsPeerstore/Badger/AddGetAndClearAddrs-10Addrs-Caching-8             427            427            +0.00%
BenchmarkDsPeerstore/Badger/AddGetAndClearAddrs-100Addrs-Caching-8            2481           2441           -1.61%
BenchmarkDsPeerstore/Badger/Get1000PeersWithAddrs-1Addrs-Caching-8            18140          18140          +0.00%
BenchmarkDsPeerstore/Badger/Get1000PeersWithAddrs-10Addrs-Caching-8           18141          18140          -0.01%
BenchmarkDsPeerstore/Badger/Get1000PeersWithAddrs-100Addrs-Caching-8          18352          18352          +0.00%
BenchmarkDsPeerstore/Badger/GetAddrs-1Addrs-Caching-8                         2              2              +0.00%
BenchmarkDsPeerstore/Badger/GetAddrs-10Addrs-Caching-8                        2              2              +0.00%
BenchmarkDsPeerstore/Badger/GetAddrs-100Addrs-Caching-8                       2              2              +0.00%
BenchmarkDsPeerstore/Badger/SetAddrs-1Addrs-Caching-8                         137            138            +0.73%
BenchmarkDsPeerstore/Badger/SetAddrs-10Addrs-Caching-8                        350            350            +0.00%
BenchmarkDsPeerstore/Badger/SetAddrs-100Addrs-Caching-8                       2380           2380           +0.00%
BenchmarkDsPeerstore/Badger#01/AddAddrs-1Addrs-Cacheless-8                    134            134            +0.00%
BenchmarkDsPeerstore/Badger#01/AddAddrs-10Addrs-Cacheless-8                   347            347            +0.00%
BenchmarkDsPeerstore/Badger#01/AddAddrs-100Addrs-Cacheless-8                  2376           2376           +0.00%
BenchmarkDsPeerstore/Badger#01/AddGetAndClearAddrs-1Addrs-Cacheless-8         237            237            +0.00%
BenchmarkDsPeerstore/Badger#01/AddGetAndClearAddrs-10Addrs-Cacheless-8        490            490            +0.00%
BenchmarkDsPeerstore/Badger#01/AddGetAndClearAddrs-100Addrs-Cacheless-8       2926           2926           +0.00%
BenchmarkDsPeerstore/Badger#01/Get1000PeersWithAddrs-1Addrs-Cacheless-8       18140          18140          +0.00%
BenchmarkDsPeerstore/Badger#01/Get1000PeersWithAddrs-10Addrs-Cacheless-8      18141          18141          +0.00%
BenchmarkDsPeerstore/Badger#01/Get1000PeersWithAddrs-100Addrs-Cacheless-8     18352          18352          +0.00%
BenchmarkDsPeerstore/Badger#01/GetAddrs-1Addrs-Cacheless-8                    28             28             +0.00%
BenchmarkDsPeerstore/Badger#01/GetAddrs-10Addrs-Cacheless-8                   68             68             +0.00%
BenchmarkDsPeerstore/Badger#01/GetAddrs-100Addrs-Cacheless-8                  431            431            +0.00%
BenchmarkDsPeerstore/Badger#01/SetAddrs-1Addrs-Cacheless-8                    134            134            +0.00%
BenchmarkDsPeerstore/Badger#01/SetAddrs-10Addrs-Cacheless-8                   347            347            +0.00%
BenchmarkDsPeerstore/Badger#01/SetAddrs-100Addrs-Cacheless-8                  2384           2376           -0.34%

benchmark                                                                     old bytes     new bytes     delta
BenchmarkDsPeerstore/Badger/AddAddrs-1Addrs-Caching-8                         11336         10841         -4.37%
BenchmarkDsPeerstore/Badger/AddAddrs-10Addrs-Caching-8                        22491         22490         -0.00%
BenchmarkDsPeerstore/Badger/AddAddrs-100Addrs-Caching-8                       136117        136186        +0.05%
BenchmarkDsPeerstore/Badger/AddGetAndClearAddrs-1Addrs-Caching-8              13207         13205         -0.02%
BenchmarkDsPeerstore/Badger/AddGetAndClearAddrs-10Addrs-Caching-8             24981         24979         -0.01%
BenchmarkDsPeerstore/Badger/AddGetAndClearAddrs-100Addrs-Caching-8            141294        139119        -1.54%
BenchmarkDsPeerstore/Badger/Get1000PeersWithAddrs-1Addrs-Caching-8            1244610       1244609       -0.00%
BenchmarkDsPeerstore/Badger/Get1000PeersWithAddrs-10Addrs-Caching-8           1244646       1244639       -0.00%
BenchmarkDsPeerstore/Badger/Get1000PeersWithAddrs-100Addrs-Caching-8          1256096       1256100       +0.00%
BenchmarkDsPeerstore/Badger/GetAddrs-1Addrs-Caching-8                         32            32            +0.00%
BenchmarkDsPeerstore/Badger/GetAddrs-10Addrs-Caching-8                        176           176           +0.00%
BenchmarkDsPeerstore/Badger/GetAddrs-100Addrs-Caching-8                       1808          1808          +0.00%
BenchmarkDsPeerstore/Badger/SetAddrs-1Addrs-Caching-8                         10859         10841         -0.17%
BenchmarkDsPeerstore/Badger/SetAddrs-10Addrs-Caching-8                        22491         22508         +0.08%
BenchmarkDsPeerstore/Badger/SetAddrs-100Addrs-Caching-8                       135847        135845        -0.00%
BenchmarkDsPeerstore/Badger#01/AddAddrs-1Addrs-Cacheless-8                    10679         10678         -0.01%
BenchmarkDsPeerstore/Badger#01/AddAddrs-10Addrs-Cacheless-8                   22310         22310         +0.00%
BenchmarkDsPeerstore/Badger#01/AddAddrs-100Addrs-Cacheless-8                  135655        135649        -0.00%
BenchmarkDsPeerstore/Badger#01/AddGetAndClearAddrs-1Addrs-Cacheless-8         14400         14398         -0.01%
BenchmarkDsPeerstore/Badger#01/AddGetAndClearAddrs-10Addrs-Cacheless-8        28592         28590         -0.01%
BenchmarkDsPeerstore/Badger#01/AddGetAndClearAddrs-100Addrs-Cacheless-8       169664        169657        -0.00%
BenchmarkDsPeerstore/Badger#01/Get1000PeersWithAddrs-1Addrs-Cacheless-8       1244615       1244593       -0.00%
BenchmarkDsPeerstore/Badger#01/Get1000PeersWithAddrs-10Addrs-Cacheless-8      1244666       1244695       +0.00%
BenchmarkDsPeerstore/Badger#01/Get1000PeersWithAddrs-100Addrs-Cacheless-8     1256083       1256127       +0.00%
BenchmarkDsPeerstore/Badger#01/GetAddrs-1Addrs-Cacheless-8                    1360          1360          +0.00%
BenchmarkDsPeerstore/Badger#01/GetAddrs-10Addrs-Cacheless-8                   3920          3920          +0.00%
BenchmarkDsPeerstore/Badger#01/GetAddrs-100Addrs-Cacheless-8                  29264         29264         +0.00%
BenchmarkDsPeerstore/Badger#01/SetAddrs-1Addrs-Cacheless-8                    10679         10678         -0.01%
BenchmarkDsPeerstore/Badger#01/SetAddrs-10Addrs-Cacheless-8                   22312         22310         -0.01%
BenchmarkDsPeerstore/Badger#01/SetAddrs-100Addrs-Cacheless-8                  136071        135651        -0.31%
```

Reply:
Not using transactions in write operations doesn't hurt consistency, as we load the entry in memory and lock it before changing it (read-through approach). However, I need to find a way to avoid dirty reads during GC. If an entry is modified while it's being GC'ed (e.g. by a delete, update, etc.), we could end up with a lost update.

Reply:
@Stebalien regarding the `ds.Reader` and `ds.Writer` interfaces, YES. This would be very convenient. I'll define local interfaces for now (really keen to see this merged), and then I can visit the reader/writer topic.

Reply:
go-datastore now exposes these interfaces, and I've adopted `ds.Write` in this commit: https://github.com/libp2p/go-libp2p-peerstore/pull/47/commits/20a0dc649da208224a47cd67f7bf8e540d1b9b64.